src/wallet/sendTransaction.ts

```diff
@@ -0,0 +1,353 @@
+      maxTxMiningRetries: number,
+    };
+    const newOptions: mineOptionsType = Object.assign({
+      startMiningTx: true,
+      maxTxMiningRetries: 3,
+    }, options);
+
+    this.mineTransaction = new MineTransaction(this.transaction, { maxTxMiningRetries: newOptions.maxTxMiningRetries });
+
+    this.mineTransaction.on('mining-started', () => {
+      this.emit('mine-tx-started');
+    });
+
+    this.mineTransaction.on('estimation-updated', (data) => {
+      this.emit('estimation-updated', data);
+    })
+
+    this.mineTransaction.on('job-submitted', (data) => {
+      this.emit('job-submitted', data);
+    })
+
+    this.mineTransaction.on('job-done', (data) => {
+      this.emit('job-done', data);
+    })
+
+    this.mineTransaction.on('error', (message) => {
+      this.emit('send-error', message);
+    })
+
+    this.mineTransaction.on('unexpected-error', (message) => {
+      this.emit('send-error', message);
+    })
+
+    this.mineTransaction.on('success', (data) => {
+      this.transaction!.parents = data.parents;
+      this.transaction!.timestamp = data.timestamp;
+      this.transaction!.nonce = data.nonce;
+      this.transaction!.weight = data.weight;
+      this.emit('mine-tx-ended', data);
+    })
+
+    if (newOptions.startMiningTx) {
+      this.mineTransaction.start();
+    }
+
+    return this.mineTransaction;
+  }
+
+  /**
+   * Create and send a tx proposal to wallet service
+   * Expects this.transaction to be prepared, signed and mined
+   *
+   * @memberof SendTransactionWalletService
+   * @inner
+   */
+  async handleSendTxProposal() {
+    if (this.transaction === null) {
+      throw new WalletError('Can\'t push transaction if it\'s null.');
+    }
+    this.emit('send-tx-start', this.transaction);
+    const txHex = this.transaction.toHex();
+
+    try {
+      const responseData = await walletApi.createTxProposal(this.wallet, txHex);
+      const txProposalId = responseData.txProposalId;
+      const sendData = await walletApi.updateTxProposal(this.wallet, txProposalId, txHex);
+      this.transaction.updateHash();
+      this.emit('send-tx-success', this.transaction);
+    } catch (err) {
+      if (err instanceof WalletRequestError) {
+        this.emit('send-error', 'Error sending tx proposal.');
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  /**
+   * Run sendTransaction from mining, i.e. expects this.transaction to be prepared and signed
+   * then it will mine and handle tx proposal
+   *
+   * 'until' parameter can be 'mine-tx', in order to only mine the transaction without propagating
+   *
+   * @memberof SendTransactionWalletService
+   * @inner
+   */
+  runFromMining(until = null) {
+    try {
+      this.mineTx();
+      if (until === 'mine-tx') {
+        return;
+      }
+
+      this.once('mine-tx-ended', (data) => {
+        this.handleSendTxProposal();
+      });
+    } catch (err) {
+      if (err instanceof WalletError) {
+        this.emit('send-error', err);
```

Comment:
Should we omit the exception?

Reply:
The same as in this comment here: https://github.com/HathorNetwork/hathor-wallet-lib/pull/221#discussion_r687262959